### PR TITLE
Initial product form field store

### PR DESCRIPTION
**Pull request by louwie17**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

This is a very barebones version of the initial field store/registry, that will allow us to register fields that will be available through the ProductForm fields API, this will also allow plugins to render their own fields to make use off.
I mimicked some of the design implementation from [registering new Gutenberg Blocks](https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/README.md), that way this stays somewhat consistent.
Again the above is very barebones for now and will likely be expanded on quite a bit as new requirements are needed.

Partially addresses #36074  .

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

### How to test the changes in this Pull Request:

1. Run storybook - `pnpm --filter=storybook storybook`
2. Go to the `product-fields` example -> `WooCommerce Admin/experimental/product-fields`
3. Change between the two fields that are registered and make sure they are updated correctly (again pretty simple)
4. Feel free to edit the storybook as well and add some more visually different fields.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: packages/js/components/src/product-fields/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# Product Fields
+
+Product Fields are used within the WooCommerce Admin product editor, for rendering new fields using PHP.
+
+## Example
+
+```js
+// block.js
```
Comment (joelclimbsthings) [inline packages/js/components/src/product-fields/README.md]: Is this `block.js` comment from gutenberg?

File: packages/js/components/src/product-fields/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+( function ( element ) {
+	var el = element.createElement;
```
Comment (joelclimbsthings) [inline packages/js/components/src/product-fields/README.md]: Might also update this to `const`

File: packages/js/components/src/product-fields/stories/index.tsx [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	return (
+		<div>
+			<select value={ selectedField } onChange={ handleChange }>
+				{ fields.map( ( field ) => (
+					<option key={ field } value={ field }>
+						{ field }
+					</option>
+				) ) }
+			</select>
+			{ renderField( selectedField || fields[ 0 ], { name: 'test' } ) }
```
Comment (joelclimbsthings) [inline packages/js/components/src/product-fields/stories/index.tsx]: Would it be slightly simpler to just make `fields[ 0 ]` the default value of `selectedField`? Might just be my habit.

File: packages/js/components/src/product-fields/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+_Usage_
+
+```js
+import { __ } from '@wordpress/i18n';
+import { registerProductField } from '@woocommerce/components';
+
+registerProductField( 'number', {
+	name: 'number',
+	edit: () => {
```
Comment (joelclimbsthings) [inline packages/js/components/src/product-fields/README.md]: Do we want a description here somewhere of what `edit()` is meant to do?

File: packages/js/components/src/product-fields/api/registration.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * registerProductFieldType( 'attributes-field', {
+ * } );
+ * ```
+ */
+export function registerProductField(
+	fieldName: string,
+	settings: ProductFieldDefinition
+) {
+	const name = fieldName;
```
Comment (joelclimbsthings) [inline packages/js/components/src/product-fields/api/registration.ts]: Is there a reason why we're creating another `name` variable here?

File: packages/js/components/src/product-fields/api/registration.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+	if ( select( productFieldStore ).getProductField( name ) ) {
+		// eslint-disable-next-line no-console
+		console.error( 'Product Field "' + name + '" is already registered.' );
+		return;
+	}
+
+	const blockType = {
```
Comment (joelclimbsthings) [inline packages/js/components/src/product-fields/api/registration.ts]: Is this reference to `block` from gutenberg? Also, we don't really need the variable here; could just insert the object below where it's used.

**Review by joelclimbsthings — DISMISSED** [review]
Great work @louwie17 ! This is working well in this limited test. I left a few minor comments. I also had an issue running storybook initially since I was missing the `rememo` dependency; is that something you meant to add?

**Comment by louwie17** [comment]
@joelclimbsthings thanks for the feedback, I addressed all your suggestions.
I also end up using `memoize-one` over `rememo` as that was already part of the component package.json and they are practically the same thing.
I didn't necessarily want to install a new package if we plan on moving these out in the near future (potentially).

**Comment by joelclimbsthings** [comment]
Great, thanks @louwie17 . Looks good! :shipit:

**Review by joelclimbsthings — APPROVED** [review]

**Pull request merged**